Source/WebCore/Modules/mediastream/VideoTrackGenerator.cpp

```diff
@@ -122,13 +125,34 @@ Ref<MediaStreamTrack> VideoTrackGenerator::track()
     return Ref { m_track };
 }
 
-VideoTrackGenerator::Source::Source()
+VideoTrackGenerator::Source::Source(ScriptExecutionContextIdentifier identifier)
     : RealtimeMediaSource(CaptureDevice { { }, CaptureDevice::DeviceType::Camera, emptyString() })
+    , m_contextIdentifier(identifier)
 {
 }
 
+void VideoTrackGenerator::Source::endProducingData()
+{
+    ASSERT(isMainThread());
+    ScriptExecutionContext::postTaskTo(m_contextIdentifier, [weakThis = ThreadSafeWeakPtr { *this }] (auto&) {
+        auto protectedSource = weakThis.get();
+        RefPtr writable = protectedSource ? protectedSource->m_writable.get() : nullptr;
+        if (writable)
```

Comment:
this seems racy at a glance, we have m_writable written on the main thread but read on the event loop with no memory barriers (locks) in place.

If this is all main thread, can we add the `WTF_GUARDED_BY_CAPABILITY(mainThread)` required annotation for the m_writable member?

Will need to change the `ASSERT(isMainThread())` for `assertIsMainThread()`

Reply:
m_writable is only accessed in the worker thread (see the setter below), VideoTrackGenerator is only created in worker contexts.
The call to ScriptExecutionContext::postTaskTo makes sure we hop to the right thread to manipulate m_writable.
Also, m_writable is a RefCounted object so refing/derefing it in multiple threads should already be covered by asserts.

Reply:
My bad, I misread the code, I missed the `!`